dartsim/src/FreeGroupFeatures.cc

```diff
@@ -95,9 +119,31 @@ FreeGroupFeatures::FreeGroupInfo FreeGroupFeatures::GetCanonicalInfo(
   const auto model_it = this->models.idToObject.find(_groupID);
   if (model_it != this->models.idToObject.end())
   {
-    return FreeGroupInfo{
-      model_it->second->model->getRootBodyNode(),
-      model_it->second->model.get()};
+    const auto &modelInfo = model_it->second;
+    if (modelInfo->model->getNumBodyNodes() > 0)
+    {
+      return FreeGroupInfo{
+        modelInfo->model->getRootBodyNode(),
+          modelInfo->model.get()};
+    }
+    else
+    {
+      // If the skeleton doesn't have any BodyNodes, we recursively search for
+      // a root BodyNode in the first nested model that has a non-zero number of
+      // BodyNodes. Since all root joints have been verified to be FreeJoints in
+      // FindFreeGroupForModel, we are guaranteed that the BodyNode found in
+      // this way is the root link for the free group.
+      for (const auto &nestedModel : modelInfo->nestedModels)
+      {
+        auto freeGroupInfo =
+          this->GetCanonicalInfo(this->GenerateIdentity(nestedModel));
+        if (freeGroupInfo.model != nullptr || freeGroupInfo.link != nullptr)
```

Comment:
do both of these logical statements need to be true? (i.e. should this be an `&&` operator instead of `||`)

Reply:
No, when the freegroup is for a link, the model member will be `nullptr`. But I guess, I can just check for `freeGroupInfo.link` since that should never be `nullptr`.

Reply:
Changed to checking for just the link in https://github.com/ignitionrobotics/ign-physics/pull/231/commits/74b3f6d75d66eef7fb9af2caf5697e1629cb5046.

Reply:
👍